common/cloud_builder.py

```diff
@@ -0,0 +1,261 @@
+"""CloudBuilder executes agents in Cloud Builds."""
+import argparse
+import logging
+import os
+import subprocess
+import tempfile
+import time
+import uuid
+from typing import Any
+
+import google.api_core.client_options
+import googleapiclient.errors
+from google.api_core.exceptions import NotFound
+from google.auth import default
+from google.auth.transport.requests import Request
+from google.cloud import storage
+from googleapiclient.discovery import build as cloud_build
+
+import utils
+from agent.base_agent import BaseAgent
+from results import BuildResult, Result
+
+OF_REPO = 'https://github.com/google/oss-fuzz.git'
+US_CENTRAL_CLIENT_OPTIONS = google.api_core.client_options.ClientOptions(
+    api_endpoint='https://us-central1-cloudbuild.googleapis.com/')
+
+
+class CloudBuilder:
+  """A worker to execute llm-agents workflow in Google Cloud Build, providing a
+  scalable and distributed alternative to local executions:
+  - Request, monitor, and manage Google Cloud Build jobs.
+  - Execute agent in the cloud environment, replicating the local conditions.
+  - Transfer data and results between local and cloud environment.
+  """
+
+  def __init__(self, args: argparse.Namespace) -> None:
+    #TODO(dongge): extra tags.
+    self.credentials, self.project_id = default()
+    assert self.project_id, 'Cloud experiment requires a Google cloud project.'
+    assert hasattr(
+        self.credentials,
+        'refresh'), ('Cloud experiment requires a service account email')
+    assert hasattr(self.credentials, 'service_account_email'), (
+        'Cloud experiment requires a service account email')
+
+    try:
+      # TODO(dongge): Understand why this crashes in local experiments.
+      self.credentials.refresh(Request())  # type: ignore
+    except:
+      pass
+    self.bucket_name = args.cloud_experiment_bucket
+    self.bucket = storage.Client().bucket(self.bucket_name)
+
+    # pylint: disable=no-member
+    self.builds = cloud_build(
+        'cloudbuild',
+        'v1',
+        credentials=self.credentials,
+        cache_discovery=False,
+        client_options=US_CENTRAL_CLIENT_OPTIONS).projects().builds()
+    self.storage_client = storage.Client(credentials=self.credentials)
+
+  def _upload_to_gcs(self, local_file_path: str) -> str:
+    """Uploads a file to Google Cloud Storage."""
+    dest_file_name = os.path.basename(local_file_path)
+    self.bucket.blob(dest_file_name).upload_from_filename(local_file_path)
+    bucket_file_url = f'gs://{self.bucket_name}/{dest_file_name}'
+    logging.info('Uploaded %s to %s', local_file_path, bucket_file_url)
+    return bucket_file_url
+
+  def _prepare_and_upload_archive(self) -> str:
+    """Archives and uploads local OFG repo to cloud build."""
+    ofg_repo = subprocess.check_output(['git', 'rev-parse', '--show-toplevel'],
```

Comment:
Instead of shelling out to git for this, can we instead add this to `utils.py`: 

```
OFG_ROOT_DIR = os.path.abspath(os.path.dirname(os.path.dirname(__file__)))
```

(or something similar?) 

This also makes it more robust because it doesn't depend on the CWD.